### PR TITLE
Fix PP sampled token broadcast for speculative decoding

### PR DESCRIPTION
**Pull request by Debasish-87**
## Summary

This PR addresses **Failure 2** reported in #49355.

During speculative decoding with Pipeline Parallelism (PP > 1), the pipeline-parallel communication path broadcasts the sampled token IDs required by downstream pipeline stages.

Previously, the broadcast used `sampler_output.sampled_token_ids`, which represents the sampler output for the current decoding step. However, the PP communication path is designed to exchange the previous sampled token stored in `input_batch.prev_sampled_token_ids`.

As a result, the sender and receiver operate on different tensor contracts, leading to failures during PP execution.

Example error:

```text id="a4tt6m"
PP+async expects sampled_token_ids to have shape [num_reqs, 1]
```

## Root Cause

The PP receive path expects a tensor containing the previous sampled token for each request:

```python id="aykr8n"
input_batch.prev_sampled_token_ids
```

The sender, however, was broadcasting:

```python id="m6zc4r"
sampler_output.sampled_token_ids
```

These tensors have different semantics and may have different shapes depending on the speculative decoding method.

Additionally, the broadcast occurred before the bookkeeping stage that updates `input_batch.prev_sampled_token_ids`, meaning the required PP state was not yet available.

## Changes

This PR restores the intended PP communication contract by:

* broadcasting `input_batch.prev_sampled_token_ids` instead of `sampler_output.sampled_token_ids`,
* moving the broadcast to occur after the bookkeeping stage where `prev_sampled_token_ids` has been updated,
* preserving the existing PP broadcast conditions (`PP > 1`, last PP rank, and `broadcast_pp_output == False`).

With these changes, the sender transmits the same tensor that downstream PP stages already expect.

## Scope

This PR addresses **Failure 2** reported in #49355.

The remaining issues described in the same report are intentionally handled separately:

* **Failure 1:** Speculative drafter access on non-final PP ranks (**addressed in #49442**).
* **Failure 3:** CUDA device-side assert during logits indexing while running MTP speculative decoding with PP > 1 (**still under investigation**).

No changes are made to speculative decoding algorithms, scheduling logic, hidden-state computation, or pipeline execution.

## Testing

Verified locally that the PP communication path now broadcasts `input_batch.prev_sampled_token_ids` after it has been updated, matching the tensor contract expected by the PP receive path.

## Related Work

This is the second of two independent fixes for #49355.

* **#49442:** Fix speculative drafter access on non-final pipeline parallel ranks (Failure 1).
* **#49443 (this PR):** Fix PP sampled token broadcast for speculative decoding (Failure 2).

Together, these PRs address the first two failures reported in #49355. Failure 3 remains under investigation.

**Related Issue**

Related to #49355.

### CODE REVIEW

**Comment by njhill** [comment]
Please use V2 model runner `VLLM_USE_V2_MODEL_RUNNER=1`

**Comment by Debasish-87** [comment]
Thanks for the suggestion. I originally reproduced and validated this using the V1 model runner.

I'll re-run the reproduction with `VLLM_USE_V2_MODEL_RUNNER=1` and verify whether the same issue and fix apply there. I'll update this PR with the results.

**Comment by Debasish-87** [comment]
Quick update: I don't have access to a CUDA-enabled GPU locally, so I'm currently unable to validate this with `VLLM_USE_V2_MODEL_RUNNER=1`.

If there's a recommended GPU CI job or environment for validating the V2 model runner, I'm happy to use that and update the PR with the results.
